rskj-core/src/main/java/co/rsk/net/SyncProcessor.java

```diff
@@ -0,0 +1,319 @@
+package co.rsk.net;
+
+import co.rsk.core.bc.BlockChainStatus;
+import co.rsk.net.messages.*;
+import co.rsk.net.sync.*;
+import co.rsk.validators.BlockHeaderValidationRule;
+import com.google.common.annotations.VisibleForTesting;
+import org.ethereum.core.Block;
+import org.ethereum.core.BlockHeader;
+import org.ethereum.core.BlockIdentifier;
+import org.ethereum.core.Blockchain;
+import org.ethereum.crypto.HashUtil;
+import org.ethereum.util.ByteUtil;
+import org.ethereum.validator.DependentBlockHeaderRule;
+import org.ethereum.validator.DifficultyRule;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.time.Duration;
+import java.util.*;
+
+/**
+ * Created by ajlopez on 29/08/2017.
+ * This class' methods are executed one at a time because NodeMessageHandler is synchronized.
+ */
+public class SyncProcessor implements SyncEventsHandler {
+    private static final Logger logger = LoggerFactory.getLogger("syncprocessor");
+
+    private final SyncConfiguration syncConfiguration;
+
+    private Blockchain blockchain;
+    private BlockSyncService blockSyncService;
+    private PeersInformation peerStatuses;
+
+    private SyncState syncState;
+    private NodeID selectedPeerId;
+    private PendingMessages pendingMessages;
+    private SyncInformationImpl syncInformation;
+
+    public SyncProcessor(Blockchain blockchain, BlockSyncService blockSyncService, SyncConfiguration syncConfiguration, BlockHeaderValidationRule blockHeaderValidationRule) {
+        // TODO(mc) implement FollowBestChain
+        this.blockchain = blockchain;
+        this.blockSyncService = blockSyncService;
+        this.syncConfiguration = syncConfiguration;
+        this.peerStatuses = new PeersInformation(syncConfiguration);
+        this.syncInformation = new SyncInformationImpl(blockHeaderValidationRule);
+        this.pendingMessages = new PendingMessages();
+        setSyncState(new DecidingSyncState(this.syncConfiguration, this, syncInformation, peerStatuses));
+    }
+
+    public void processStatus(MessageChannel sender, Status status) {
+        logger.trace("Receiving syncState from node {} block {} {}", sender.getPeerNodeID(), status.getBestBlockNumber(), HashUtil.shortHash(status.getBestBlockHash()), status.getBestBlockHash());
+        // we keep track of best known block
+        final long peerBestBlockNumber = status.getBestBlockNumber();
+
+        if (peerBestBlockNumber > blockSyncService.getLastKnownBlockNumber())
+            blockSyncService.setLastKnownBlockNumber(peerBestBlockNumber);
+
+        this.peerStatuses.getOrRegisterPeer(sender).setStatus(status);
+        this.syncState.newPeerStatus();
+    }
+
+    public void processSkeletonResponse(MessageChannel sender, SkeletonResponseMessage message) {
+        logger.trace("Process skeleton response from node {}", sender.getPeerNodeID());
+        peerStatuses.getOrRegisterPeer(sender);

```

Comment:
If this is needed it sounds like we have an issue

Reply:
The same applies for all responses below

Reply:
I don't believe it's a problem at this place. We can discuss if we should be more punitive when unexpected messages arrive.